src/server/JSON/JSON.cpp

```diff
@@ -0,0 +1,258 @@
+std::string SerializeContainer(const Container& container, bool pretty, int indent, const std::string& open,
+							   const std::string& close, Formatter formatter)
+{
+	std::string indent_str(indent, ' ');
+	std::string result = open; // Add opening delimiter.
+
+	if (pretty) result += "\n";
+	size_t count = 0;
+	// Iterate through each element in the container.
+	for (const auto& elem : container)
+	{
+		if (pretty) result += std::string(indent + 2, ' '); // Extra indent for elements.
+		result += formatter(elem, indent + 2, pretty);		// Format the element.
+
+		if (++count != container.size()) result += ","; // Separate elements with a comma.
+		if (pretty) result += "\n";
+	}
+	if (pretty) result += indent_str; // Matching indent for closing delimiter.
+	result += close;				  // Add closing delimiter.
+	return result;
+}
+} // namespace
+
+// Serialize converts the JSON value into its JSON-formatted string.
+std::string JSON::Serialize(bool pretty, int indent) const
+{
+	switch (GetType())
+	{
+	case Type::NUL:
+		return "null";
+
+	case Type::BOOL:
+		return AsBool() ? "true" : "false";
+
+	case Type::NUMBER:
+	{
+		double num = AsNumber();
+		double intpart;
+		std::ostringstream oss;
+		// Check if the number has no fractional part.
+		if (std::modf(num, &intpart) == 0.0)
+			oss << static_cast<long long>(intpart);
+		else
+			oss << num;
+		return oss.str();
+	}
+
+	case Type::STRING:
+		// Enclose the escaped string in double quotes.
+		return "\"" + EscapeString(AsString()) + "\"";
+
+	case Type::ARRAY:
+		// Helper to serialize an array.
+		return SerializeContainer(AsArray(), pretty, indent, "[", "]", [](const JSON& elem, int newIndent, bool pretty)
+								  { return elem.Serialize(pretty, newIndent); });
+
+	case Type::OBJECT:
+		// Helper to serialize an object.
+		return SerializeContainer(AsObject(), pretty, indent, "{", "}",
+								  [](const std::pair<const std::string, JSON>& p, int newIndent, bool pretty)
+								  {
+									  std::string res = "\"" + EscapeString(p.first) + "\":";
+									  if (pretty) res += " ";
+									  res += p.second.Serialize(pretty, newIndent);
+									  return res;
+								  });
+
+	default:
+		throw std::runtime_error("Unknown JSON type in serialization");
+	}
+}
+
+std::string JSON::EscapeString(const std::string& input)
+{
+	// Local lambda to escape a single character.
+	auto escape_char = [](char c) -> std::string
+	{
+		switch (c)
+		{
+		case '\"':
+			return "\\\"";
+		case '\\':
+			return "\\\\";
+		case '\b':
+			return "\\b";
+		case '\f':
+			return "\\f";
+		case '\n':
+			return "\\n";
+		case '\r':
+			return "\\r";
+		case '\t':
+			return "\\t";
+		default:
+			if (static_cast<unsigned char>(c) < 0x20 || static_cast<unsigned char>(c) > 0x7E)
+			{
+				std::ostringstream oss;
+				oss << "\\u" << std::hex << std::setw(4) << std::setfill('0') << (int)c;
+				return oss.str();
+			}
+			else
+			{
+				return {c};
+			}
```

Comment:
What if we try to make it more readable by separating some code into small functions:
```cpp
// Check if the character is non-printable (outside the printable ASCII range)
bool isNonPrintableChar(char c) {
    unsigned char uc = static_cast<unsigned char>(c);
    return uc < 0x20 || uc > 0x7E;
}

// Convert a non-printable character to a Unicode escape sequence
std::string convertToUnicodeEscape(char c) {
    std::ostringstream oss;
    unsigned char uc = static_cast<unsigned char>(c);
    oss << "\\u" << std::hex << std::setw(4) << std::setfill('0') << static_cast<int>(uc);
    return oss.str();
}
```
```suggestion
                        return isNonPrintableChar(c) ? convertToUnicodeEscape(c) : {c};
```